protocol-units/settlement/mcr/client/src/eth_client.rs

```diff
@@ -0,0 +1,536 @@
+		let provider_client1 = ProviderBuilder::new()
+			.with_recommended_fillers()
+			.signer(EthereumSigner::from(signer1))
+			.on_http(rpc_url.parse().unwrap());
+
+		let config = McrEthSettlementConfig {
+			mrc_contract_address: mcr_address.to_string(),
+			gas_limit: DEFAULT_TX_GAS_LIMIT,
+			tx_send_nb_retry: MAX_TX_SEND_RETRY,
+		};
+
+		let client1 = McrEthSettlementClient::build_with_provider(
+			provider_client1,
+			signer1_addr,
+			ws_url.clone(),
+			config.clone(),
+		)
+		.await
+		.unwrap();
+
+		let mut client1_stream = client1.stream_block_commitments().await.unwrap();
+
+		//client post a new commitment
+		let commitment =
+			BlockCommitment { height: 1, block_id: Id([2; 32]), commitment: Commitment([3; 32]) };
+
+		let res = client1.post_block_commitment(commitment.clone()).await;
+		assert!(res.is_ok());
+
+		//no notification quorum is not reach
+		let res =
+			tokio::time::timeout(tokio::time::Duration::from_secs(5), client1_stream.next()).await;
+		assert!(res.is_err());
+
+		//Build client 2 and send the second commitment.
+		let client2 =
+			McrEthSettlementClient::build_with_urls(&rpc_url, ws_url, &anvil_address[2].1, config)
+				.await
+				.unwrap();
+
+		let mut client2_stream = client2.stream_block_commitments().await.unwrap();
+
+		//client post a new commitment
+		let res = client2.post_block_commitment(commitment).await;
+		assert!(res.is_ok());
+
+		// now we move to block 2 and make some commitment just to trigger the epochRollover
+		let commitment2 =
+			BlockCommitment { height: 2, block_id: Id([4; 32]), commitment: Commitment([5; 32]) };
+
+		let res = client2.post_block_commitment(commitment2.clone()).await;
+		assert!(res.is_ok());
+
+		//validate that the accept commitment stream get the event.
+		let event =
+			tokio::time::timeout(tokio::time::Duration::from_secs(5), client1_stream.next())
+				.await
+				.unwrap()
+				.unwrap()
+				.unwrap();
+		assert_eq!(event.commitment.0[0], 3);
+		assert_eq!(event.block_id.0[0], 2);
+		let event =
+			tokio::time::timeout(tokio::time::Duration::from_secs(5), client2_stream.next())
+				.await
+				.unwrap()
+				.unwrap()
+				.unwrap();
+		assert_eq!(event.commitment.0[0], 3);
+		assert_eq!(event.block_id.0[0], 2);
+
+		//test post batch commitment
+		// post the complementary batch on height 2 and one on height 3
+		let commitment3 =
+			BlockCommitment { height: 3, block_id: Id([6; 32]), commitment: Commitment([7; 32]) };
+		let res = client1.post_block_commitment_batch(vec![commitment2, commitment3]).await;
+		assert!(res.is_ok());
+		//validate that the accept commitment stream get the event.
+		let event =
+			tokio::time::timeout(tokio::time::Duration::from_secs(5), client1_stream.next())
+				.await
+				.unwrap()
+				.unwrap()
+				.unwrap();
+		assert_eq!(event.commitment.0[0], 5);
+		assert_eq!(event.block_id.0[0], 4);
+		let event =
+			tokio::time::timeout(tokio::time::Duration::from_secs(5), client2_stream.next())
```

Comment:
Even in tests, I would prefer not use unwrap as it just makes everything more verbose.

Reply:
I'm mostly agreeing with you, but I found that unwrap failure is more informative that using a ?.
For example, when getting an environment variable with unwrap:
```
---- eth_client::test::test_send_commitment stdout ----
thread 'eth_client::test::test_send_commitment' panicked at protocol-units/settlement/mcr/client/src/eth_client.rs:294:52:
called `Result::unwrap()` on an `Err` value: NotPresent
```
And with ?
```
failures:

---- eth_client::test::test_send_commitment stdout ----
Error: environment variable not found
```
To find the issue, it's easier when you have the line in the file that why I keep them when you can have, or I add issues.
Perhaps we can format the `?` output in a more informative way. Using RUST_BACKTRACE=1 you have all the stack trace, but with Tokio it's not always easy to find the right root place.
 It's like you prefer, I can replace with `???`